src/browser/wisp_network.js

```diff
@@ -0,0 +1,571 @@
+        wispws.onmessage = (event) => {
+            processIncomingWispFrame(new Uint8Array(event.data));
+        };
+        wispws.onerror = () => {
+            registerWS();
+        };
+        wispws.onclose = () => {
+            registerWS();
+        };
+    };
+    registerWS();
+    config = config || {};
+    this.bus = bus;
+    this.id = config.id || 0;
+    this.router_mac = new Uint8Array((config.router_mac || "52:54:0:1:2:3").split(":").map(function(x) { return parseInt(x, 16); }));
+    this.router_ip = new Uint8Array((config.router_ip || "192.168.86.1").split(".").map(function(x) { return parseInt(x, 10); }));
+    this.vm_ip = new Uint8Array((config.vm_ip || "192.168.86.100").split(".").map(function(x) { return parseInt(x, 10); }));
+    this.masquerade = config.masquerade === undefined || !!config.masquerade;
+    this.vm_mac = new Uint8Array(6);
+
+    this.tcp_conn = {};
+
+    this.bus.register("net" + this.id + "-mac", function(mac) {
+        this.vm_mac = new Uint8Array(mac.split(":").map(function(x) { return parseInt(x, 16); }));
+    }, this);
+    this.bus.register("net" + this.id + "-send", function(data)
+    {
+        this.send(data);
+    }, this);
+}
+
+WispNetworkAdapter.prototype.destroy = function()
+{
+};
+
+// https://stackoverflow.com/questions/4460586/javascript-regular-expression-to-check-for-ip-addresses
+function validateIPaddress(ipaddress) {
+    if(/^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(ipaddress)) {
+      return true;
+    }
+    return false;
+}
+
+// DNS over HTTPS fetch, recursively fetch the A record until the first result is an IPv4
+async function dohdns(q) {
+    const prefered_fetch = (window.anura?.net?.fetch) || fetch;
+    const req = await prefered_fetch(`https://dns.google/resolve?name=${q.name.join(".")}&type=${q.type}`);
+    if(req.status === 200) {
+        const res = await req.json();
+        if(res.Answer) {
+            if(validateIPaddress(res.Answer[0].data)) {
+                return res;
+            } else {
+                return await dohdns({name: res.Answer[0].data.split("."), type: q.type});
+            }
+        }
+        return { // if theres an error, Naively return localhost
+            "Status": 0,
+            "TC": false,
+            "RD": true,
+            "RA": true,
+            "AD": true,
+            "CD": false,
+            "Question": [
+                {
+                    "name": q.name.join("."),
+                    "type": 1
+                }
+            ],
+            "Answer": [
+                {
+                    "name": q.name.join("."),
+                    "type": 1,
+                    "TTL": 600,
+                    "data": "127.0.0.1"
+                }
+            ]
+        };
+    } else {
+        throw new Error("DNS Server returned error code");
```

Comment:
Probably better to write `dbg_log("DNS Server returned error code: " + e, LOG_NET);`